himan-scripts/icing-base-top.lua

```diff
@@ -0,0 +1,155 @@
+-- The reasoning is that the pressures can be directly converted into flight levels (FL)
+hitool:SetHeightUnit(HPParameterUnit.kHPa)
+
+-- Get surface pressure
+local p = luatool:Fetch(current_time, level(HPLevelType.kHeight, 0), param("P-PA"), current_forecast_type)
+
+function BaseHPa(threshold)
+  -- Find the base of icing define as the lowest height / highest pressure at which icing index crosses the threshold value
+  -- returns the height as pressure based coordinate
+
+  local zerodata = {}
+  local pFL300data = {}
+  local thresholddata = {}
+
+  for i = 1, #p do
+     zerodata[i] = p[i] / 100
+     thresholddata[i] = threshold
+
+     -- FL300 (30000ft=301hPa)
+     pFL300data[i] = 301
+  end
+
+  local basedata = hitool:VerticalHeightGreaterThanGrid(IceParam, zerodata, pFL300data, thresholddata, 1)
+
+  return basedata
+end
+
+function BaseM(threshold)
+  -- Find the base of icing define as the lowest height / highest pressure at which icing index crosses the threshold value
+  -- returns the height as metric distance above ground
+
+  local zerodata = {}
+  local hFL300data = {}
+  local thresholddata = {}
+
+  for i = 1, #p do
+     zerodata[i] = 0
+     thresholddata[i] = threshold
+
+     -- FL300 (9144m)
+     hFL300data[i] = 9144
+  end
+
+  local basedata = hitool:VerticalHeightGreaterThanGrid(IceParam, zerodata, hFL300data, thresholddata, 1)
+
+  return basedata
+end
+
+function TopHPa(threshold,basedata)
+  -- Find the next top above the base of icing define as the height / pressure at which icing index crosses the threshold value
+  -- returns the height as pressure
+
+  local pFL300data = {}
+  local thresholddata = {}
+
+  for i = 1, #basedata do
+     thresholddata[i] = threshold
+
+     -- FL300 (30000ft=301hPa)
+     pFL300data[i] = 301
+  end
+
+  local topdata = hitool:VerticalHeightLessThanGrid(IceParam, basedata, pFL300data, thresholddata, 1)
+
+  return topdata
+end
+
+function TopM(threshold,basedata)
+  -- Find the next top above the base of icing define as the height / pressure at which icing index crosses the threshold value
+  -- returns the height as metric distance above ground
+
+  local hFL300data = {}
+  local thresholddata = {}
+
+  for i = 1, #basedata do
+     thresholddata[i] = threshold
+
+     -- FL300 (9144m)
+     hFL300data[i] = 9144
+  end
+
+  local topdata = hitool:VerticalHeightLessThanGrid(IceParam, basedata, hFL300data, thresholddata, 1)
+
+  return topdata
+end
+
+function AddScalar(arr, scalar)
+  local ret = {}
+  for i=1,#arr do
+    ret[i] = arr[i] + scalar
+  end
+  return ret
+end
+
+local baseHPa = BaseHPa(4)
+local topHPa = TopHPa(4,AddScalar(baseHPa,-1))
+
+-- Convert top [hPa] to FL
+local topFL = {}
+local baseFL = {}
+for i=1, #topHPa do
+  topFL[i] = FlightLevel_(topHPa[i] * 100)
+  baseFL[i] = FlightLevel_(baseHPa[i] * 100)
+end
+
+hitool:SetHeightUnit(HPParameterUnit.kM)
+
+local baseM = BaseM(4)
```

Comment:
Could you fetch the metric height directly when you have the barometric height? Something like this:

`local baseM = hitool:VerticalValue(param("P-PA"), baseHPa)`

Nothing wrong with current method, just an idea.